azure/services/aso/aso_test.go

```diff
@@ -0,0 +1,581 @@
+/*
+Copyright 2023 The Kubernetes Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package aso
+
+import (
+	"context"
+	"errors"
+	"testing"
+
+	asoresourcesv1 "github.com/Azure/azure-service-operator/v2/api/resources/v1api20200601"
+	"github.com/Azure/azure-service-operator/v2/pkg/genruntime"
+	"github.com/Azure/azure-service-operator/v2/pkg/genruntime/conditions"
+	"github.com/golang/mock/gomock"
+	. "github.com/onsi/gomega"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/runtime"
+	"k8s.io/apimachinery/pkg/types"
+	"k8s.io/utils/pointer"
+	"sigs.k8s.io/cluster-api-provider-azure/azure"
+	"sigs.k8s.io/cluster-api-provider-azure/azure/mock_azure"
+	gomockinternal "sigs.k8s.io/cluster-api-provider-azure/internal/test/matchers/gomock"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+	fakeclient "sigs.k8s.io/controller-runtime/pkg/client/fake"
+)
+
+type ErroringGetClient struct {
+	client.Client
+	err error
+}
+
+func (e ErroringGetClient) Get(_ context.Context, _ client.ObjectKey, _ client.Object, _ ...client.GetOption) error {
+	return e.err
+}
+
+type ErroringUpdateClient struct {
+	client.Client
+	err error
+}
+
+func (e ErroringUpdateClient) Update(_ context.Context, _ client.Object, _ ...client.UpdateOption) error {
+	return e.err
+}
+
+type ErroringDeleteClient struct {
+	client.Client
+	err error
+}
+
+func (e ErroringDeleteClient) Delete(_ context.Context, _ client.Object, _ ...client.DeleteOption) error {
+	return e.err
+}
+
+// TestCreateOrUpdateResource tests the CreateOrUpdateResource function.
+func TestCreateOrUpdateResource(t *testing.T) {
+	t.Run("ready status unknown", func(t *testing.T) {
```

Comment:
I elected not to use table tests here because even though a lot of this looks like boilerplate, just about every test does something slightly different in a way I have a hunch would be really messy to encode in a table.